[Yeluri Ketan] iP

YeluriKetan opened the pull request:
# Duke
> Personalization and customization pending as of Week 3


All you need to do is:
1.  download it from  [here](https://github.com/YeluriKetan/ip/releases/download/A-Jar/ip.jar).
2.  double-click it
3.  add your tasks.
4.  let ~~you~~ it manage your tasks for you :handshake:
###
Follow this if double-clicking the jar file doesn't work
Open up a command prompt terminal and type in the following line
```sh
java -jar (path of the executable file)
```
And it is **FREE**!!!

## Usage
The following are the commands that Duke understands
 - *todo* 'description'
 - *deadline* 'description' */by* 'yyyy mm dd hhmm'
 - *event* 'description' */from* 'yyyy mm dd hhmm' *to* 'yyyy mm dd hhmm'
 - *list*
 - *done* 'index'
 - *delete* 'index'
 - *bye*
 
 Note the above commands are case-insensitive, meaning you can go ahead and type LisT and Duke still understands you  :exploding_head:

## Future
Features:
- [x] Managing tasks, deadlines, events
- [x] Load from and save to hard disk 
- [ ] additional interactive commands (like help, time) - coming soon
- [ ] GUI - coming soon

Changes to code base:
 - Follow coding standards
 - Refactor and restructure `DukeException`
 - Optimization where applicable
 - More JUnit tests

## Suggestions
If you, the reader/reviewer finds any part of the code that can be made better or just have suggestions/advise for improvements in general, please feel free to let me know using follow-up comments.

For e.g.  Have a suggestion to make the ``findTasks`` better?
```java
public ArrayList<Task> findTasks(String content) {
    ArrayList<Task> findTasksResult = new ArrayList<>();
    for (Task currTask: tasks) {
        if (currTask.contains(content)) {
	    findTasksResult.add(currTask);
	}
    }
    return findTasksResult;
}
```

Commit:
Add Gradle support

YeluriKetan committed:
greet, echo, exit

YeluriKetan committed:
Merge pull request #1 from YeluriKetan/development

greet, echo, exit

YeluriKetan committed:
add, list, restructure

YeluriKetan committed:
added mark done

YeluriKetan committed:
todo, event, deadline

YeluriKetan committed:
Merge pull request #2 from YeluriKetan/development

merge level-2,3,4

YeluriKetan committed:
automated text ui testing

YeluriKetan committed:
Merge pull request #3 from YeluriKetan/development

automated text ui testing

YeluriKetan committed:
handle errors

YeluriKetan committed:
Merge pull request #4 from YeluriKetan/development

handle errors

YeluriKetan committed:
handle command errors

YeluriKetan committed:
Add delete command

YeluriKetan committed:
Add load and quick save abilities

YeluriKetan committed:
Add date and time functionality

YeluriKetan committed:
Merge branch 'branch-Level-7'

YeluriKetan committed:
Resolved merge conflicts from branches Level-7 and Level-8

YeluriKetan committed:
Refactor code for more OOP

YeluriKetan committed:
Divide classes into packages

YeluriKetan committed:
Add JUnit tests

YeluriKetan committed:
Add JavaDoc comments

YeluriKetan committed:
Tweak code to follow coding standard

YeluriKetan committed:
Add find command

YeluriKetan committed:
Merge branch 'branch-A-JavaDoc'

YeluriKetan committed:
Merge branch 'branch-A-CodingStandard'

YeluriKetan committed:
Merge and resolve conflicts from 'branch-Level-9'

YeluriKetan committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

YeluriKetan committed:
Add and configure gradle build automation

YeluriKetan committed:
Merge branch 'branch-A-Gradle'

YeluriKetan committed:
Fix coding standard violations

YeluriKetan committed:
Merge branch 'branch-A-CheckStyle'

jennibearduit commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,122 @@
+        LocalDateTime deadlineDT;
+
+        try {
+            deadlineDT = LocalDateTime.parse(dateTime, formatter);
```
Perhaps can consider renaming `deadlineDT` to `deadlineDt` since coding standard does not allow abbreviations to be in uppercase, or alternatively `deadlineDateTime` will be much clearer in my opinion.

naranghardik16 replied:
Not sure I like the naming of the variable <code>deadlineDT</code> as an acronym/abbreviation has been used in this variable name and the suggested way of naming such a variable should be <code>deadlineDt</code>.

jennibearduit commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,134 @@
+            break;
+        case "event":
+            LocalDateTime from = Parser.parseDateTime(parsedCommand[3]);
+            LocalDateTime to = Parser.parseDateTime(parsedCommand[3]);
```
Perhaps you can consider renaming the variable `from` into `startDateTime` to make your code clearer? Similarly, `to` can be renamed into `endDateTime`. I think the variable `from` and `to` sounds a little ambiguous to me as a reader.

naranghardik16 replied:
Not sure I like the naming of the variable <code>to</code> as it sounds a bit confusing to me. Maybe you could consider changing it to <code>endDateTime</code>.

jennibearduit commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,154 @@
+                    break;
+                case "D":
+                    tasks.add(new Deadline(nextLineArr[2], Boolean.parseBoolean(nextLineArr[1]),
+                            LocalDateTime.parse(nextLineArr[3])));
```
Perhaps array index can be stored as constants like `KEYWORD_INDEX` and `DESCRIPTION_INDEX` instead of magic numbers. For example, if I want to change the user input format sequence, I can simply change the constant value.

jennibearduit commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,154 @@
+            if (nextLine == null) {
+                break;
+            } else {
+                newTaskData.append(nextLine + "\n");
```
This line doesn't have to be inside an else block since the `if` block acts as a guard clause. This will also make it easier for reviewers to follow the conditional flow.

jennibearduit commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,136 @@
+public class Ui {
+
+    /** Line divider for beautification of command line interface */
+    private static final String H_LINE = "------------------------------\n";
```
Perhaps can consider renaming this as `LINE_BREAK` instead, since people may not understand what `H_LINE` means without more context of the app's CLI.

jennibearduit commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public String toStorage() {
+        return ("D%" + isDone + "%" + description + "%" + byDateTime + "\n");
```
Instead of using `"%"` as a magic string, perhaps keep it in a constant (maybe of `Storage`?) in any case you wish to change the delimiter symbol next time.

jennibearduit reviewed (commented):
LGTM, no major issues with code quality. Abstraction can be improved and can avoid using magic literals.

naranghardik16 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,134 @@
+            newTask = new Todo(parsedCommand[2]);
+            break;
+        case "deadline":
+            LocalDateTime by = Parser.parseDateTime(parsedCommand[3]);
```
Not sure I like the naming of the variable <code>by</code> as it sounds a bit confusing to me. Maybe you could consider changing it to <code>byDateTime</code>.

naranghardik16 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,134 @@
+            newTask = new Deadline(parsedCommand[2], by);
+            break;
+        case "event":
+            LocalDateTime from = Parser.parseDateTime(parsedCommand[3]);
```
Not sure I like the naming of the variable <code>from</code> as it sounds a bit confusing to me. Maybe you could consider changing it to <code>startDateTime</code>.

naranghardik16 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+     * @param taskIndex Index of the Task.
+     * @return Validity of the Index.
+     */
+    public boolean isIndexValid(int taskIndex) {
```
Can the variable <code>taskIndex</code> have a shorter name? I was thinking of this statement mentioned in the textbook **Variables with a large scope should have long names, variables with a small scope can have short names.**

naranghardik16 reviewed (commented):
Everything looks good to me. Can have a look at the variable naming practices and think over the suggestions provided.

YeluriKetan committed:
Add GUI capabilities

YeluriKetan committed:
Merge branch 'branch-Level-10'

YeluriKetan committed:
Add Java Assertions

There are no assertions.

Add java Assertions in select methods.

Assertions allow to verify assumptions about the program state
at various logical checkpoints during runtime.

YeluriKetan committed:
Improve code quality

Codebase doesn't follow necessary code quality.

Refactor codebase to follow code quality guidelines.

Good code quality allows readers to understand codebase
faster and better.

Code quality guidelines can be found here:
https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality

YeluriKetan committed:
Use streams in TaskList

TaskList uses loop for find operation. Hard to read
and understand operations inside loop.

Use streams with filter for find operation.

Properly written stream operations are easier to read
and understand.

YeluriKetan committed:
Merge pull request #5 from YeluriKetan/branch-A-Assertions

Add Java Assertions

YeluriKetan committed:
Merge updated branch 'master'

Missing changes from the updated master.

Merge and update this branch with the updated master.

YeluriKetan committed:
Merge updated branch 'master'

Missing changes from the updated master with asserstions.

Merge and update this branch with the updated master.

YeluriKetan committed:
Merge pull request #7 from YeluriKetan/branch-A-Streams

Use stream in Tasklist with assertions

YeluriKetan committed:
Merge updated 'master' (assertions, stream)

YeluriKetan committed:
Merge pull request #8 from YeluriKetan/branch-A-CodeQuality

Improve Code Quality (with assertions, stream)

YeluriKetan committed:
Add sort functionality

YeluriKetan committed:
Merge branch 'branch-C-Sort'

Missing sort functionality.

Add sort functionality.

Sort enables users to better view their list of tasks.

YeluriKetan committed:
Set up Continuous Integration (CI)

YeluriKetan committed:
Improve the GUI

YeluriKetan committed:
Merge branch 'branch-A-BetterGui'

YeluriKetan committed:
Refactor for personalisation of Duke

YeluriKetan committed:
Refactor codebase into better packages

YeluriKetan committed:
Fix broken test in DeadlineTest

YeluriKetan committed:
Improve code quality and add JavaDocs

YeluriKetan committed:
Add working JUnit tests

YeluriKetan committed:
Modify sort command syntax

YeluriKetan committed:
Add help command

YeluriKetan committed:
Merge branch 'development' with final changes

YeluriKetan committed:
Fix code quality issues from cs2103-bot feedback

YeluriKetan committed:
Add User Guide

YeluriKetan committed:
Set theme jekyll-theme-merlot

YeluriKetan committed:
Add project README and modify User guide

YeluriKetan committed:
Merge README changes with Github page theme update

YeluriKetan committed:
Fix User Guide for Github pages

YeluriKetan committed:
Update from branch 'master' with User Guide

YeluriKetan committed:
Add 'Case Sensitivity' to ToC in User Guide

YeluriKetan committed:
Refactor commands into separate classes

YeluriKetan committed:
Add JavaDocs for command package

YeluriKetan committed:
Fix command error scenarios and messages

YeluriKetan committed:
Refactor exceptions, Encapsulate command format info

PoseidonException is too vague,
command formats are spread over multiple classes.

Precise exceptions help pinpoint errors faster,
encapsulation for better code quality.

Create more exceptions with proper messages,
encapsulate all command formats within their respective classes.

YeluriKetan committed:
Use tab as separator value, Fix GUI font size

YeluriKetan committed:
Fix JavaDocs

YeluriKetan committed:
Add tests for command package

YeluriKetan committed:
Merge branch 'development' for improving final release

YeluriKetan committed:
Fix UserGuide and README
